Add order utility and disregarded utility calculation

nlordell opened the pull request:
In order to ultimately calculate a solution's objective value we need to calculate the utility and disregarded utility per order. This PR implements those calculations.

This PR is the first part of the PR #281 as it was too big and got difficult to review.

### Testplan

Unit tests were added with and will be run on CI to verify calculations

josojo commented on dfusion_rust_core/src/models/order.rs:
```diff
@@ -47,6 +47,73 @@ impl Order {
+    ///
+    /// Returns disregarded utility of an order or None when there was an error
+    /// in the calculation such as overflow or divide by 0.
+    pub fn disregarded_utility(
```
I think, before we merge it, we should integrate here the accounts balance, similar as it is done in the smart contract

nlordell replied:
I created a separate issue for this here #290, I referenced it in the solution objective value PR and planned on implementing it as a separate PR.

The other thing that is also missing here is update the disregarded utility calculation to include the fees, which I am waiting to see how it is done in the contracts so we do the same calculations, and get the same results.

I would rather have these in separate PRs as it would make the review a bit easier.

fleupold commented on dfusion_rust_core/src/models/order.rs:
```diff
@@ -47,6 +47,73 @@ impl Order {
+
+        // essential_utility = (exec_buy_amount - (exec_sell_amount * buy_amount)
+        //   / sell_amount) * buy_price
+        // essential_utility = (((exec_sell_amount * buy_amount) % sell_amount)
```
These comments don't match up 

`utility = essential_utility - essential_utility`

==> utility = 0

Is there any rusty way to make sure those comments are up to date? E.g. Using a documentation test or such?

fleupold commented on dfusion_rust_core/src/models/order.rs:
```diff
@@ -301,6 +368,143 @@ pub mod tests {
+            order.utility(buy_price, exec_buy_amount, exec_sell_amount),
+            // u = ((xb * os - xs * ob) * bp) / os
+            //   = ((10 * 100 - 90 * 10) * 9) / 100
+            //   = 100
```
but 100 != 9 😕

nlordell replied:
`((10 * 100 - 90 * 10) * 9) / 100 != 100` either :/. Easy fix :)

fleupold commented on dfusion_rust_core/src/models/order.rs:
```diff
@@ -47,6 +47,73 @@ impl Order {
+        let buy_amount = u128_to_u256(self.buy_amount);
+        let sell_amount = u128_to_u256(self.sell_amount);
+
+        // limit_term = sell_price * sell_amount - buy_amount * buy_price
```
Do you think this comment is necessary?

nlordell replied:
I had used them as a reference when I was writing the calculations. I guess these comments can be removed since they add no value: they are basically an almost verbatim copy of the code below them. They just confuse people when they are wrong like :point_up:.

nlordell commented:
@fleupold I addressed some of the issues you pointed out. Using doc tests sounds like a cool idea but I'm not sure exactly how we can leverage them here.

nlordell commented:
@bh2smith You Ok with me merging this? I plan on fixing the disregarded utility to account for fees and user balances in a separate PR.

bh2smith commented:
Yea boi! Merge away!
